src/lib/io/sol-gpio-contiki.c

```diff
@@ -0,0 +1,135 @@
+#include <errno.h>
+#include <stdlib.h>
+
+#include <contiki.h>
+#include <dev/button-sensor.h>
+#include <dev/leds.h>
+#include <lib/sensors.h>
+
+#define SOL_LOG_DOMAIN &_log_domain
+#include "sol-log-internal.h"
+#include "sol-gpio.h"
+#include "sol-mainloop.h"
+#include "sol-util.h"
+#include "sol-event-handler-contiki.h"
+
+static SOL_LOG_INTERNAL_DECLARE(_log_domain, "gpio");
+
+struct sol_gpio
+{
+    int pin;
+    struct sensors_sensor *button_sensor;
+    bool active_low;
+    struct {
+        void (*cb)(void *data, struct sol_gpio *gpio);
+        void *data;
```

Comment:
keep this pointer as 'const' and only cast it when calling back the cb. Instead of doing the cast when the pointer is assigned to the sol_gpio structure. This way we make it clear we should not mess with it.